### PR TITLE
Replace split-squeeze pattern

### PR DESCRIPTION
**Pull request by devashishshankar**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101767
* #101766
* __->__ #101765
* #101764

Replaces split-squeeze (same dimension) with an unbind. This will be used in combination with later patterns to remove the unbind if it follows a cat/stack

Differential Revision: [D45758181](https://our.internmc.facebook.com/intern/diff/D45758181/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/fx_passes/split_cat.py [hunk 1]
```diff
@@ -586,6 +593,89 @@ def erase_old_nodes(self, graph, split_node, next_users):
...
+        # nodes as the stored node in ctx.pattern_to_node is returned. Here we override find_anchor_nodes
+        # to not use ctx.pattern_to_node
+        for pattern in self.flat_args_kwargs[0]:
+            if isinstance(pattern, PatternExpr):
+                for other_node in pattern.find_anchor_nodes(ctx, searched):
+                    if not isinstance(other_node, torch.fx.Node):
+                        continue
+                    for node in other_node.users:
+                        if node not in searched:
+                            if not matching_nodes or self._match_fns(node):
```
Comment (jansel) [inline torch/_inductor/fx_passes/split_cat.py:611]: Can you explain what matching_nodes does?

Comment (devashishshankar) [reply torch/_inductor/fx_passes/split_cat.py:611]: If you have a better way of doing this please suggest. I was initially thinking of using a different method like `find_sibling_nodes`, but it seemed very similar to `find_anchor_nodes` in implementation.

Comment (devashishshankar) [reply torch/_inductor/fx_passes/split_cat.py:611]: btw, I just realized I forgot to add this variable to some other classes. I can add them, but waiting for your response first.

Comment (jansel) [reply torch/_inductor/fx_passes/split_cat.py:611]: This isn't really what `find_anchor_nodes()` is meant for, I worry this confuses the semantics.

I'd rather see an extra check, either in the `_match` function or in the handler that is just based on the users of split.

Comment (devashishshankar) [reply torch/_inductor/fx_passes/split_cat.py:611]: Sure, I can do that

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**